lib/scholar/metrics.ex

```diff
@@ -481,6 +482,176 @@ defmodule Scholar.Metrics do
     (diff * diff) |> Nx.mean()
   end
 
+  @doc ~S"""
+  Computes area under the curve (AUC) using the trapezoidal rule.
+
+  This is a general function, given points on a curve.
+
+  ## Examples
+
+      iex> y = Nx.tensor([0, 0, 1, 1])
+      iex> pred = Nx.tensor([0.1, 0.4, 0.35, 0.8])
+      iex> distinct_value_indices = Scholar.Metrics.calculate_distinct_value_indices(pred)
+      iex> {fpr, tpr, _thresholds} = Scholar.Metrics.roc_curve(y, pred, distinct_value_indices)
+      iex> Scholar.Metrics.auc(fpr, tpr)
+      #Nx.Tensor<
+        f32
+        0.75
+      >
+  """
+  defn auc(x, y) do
+    check_shape(x, y)
+    dx = Nx.diff(x)
+
+    # 0 means x is neither increasing nor decreasing -> error
+    direction =
+      cond do
+        Nx.all(dx <= 0) -> -1
+        Nx.all(dx >= 0) -> 1
+        true -> Nx.tensor(:nan, type: to_float_type(y))
+      end
+
+    direction * Trapezoidal.trapezoidal(y, x)
+  end
+
+  @doc ~S"""
+  It's a helper function for `Scholar.Metrics.roc_curve` and `Scholar.Metrics.roc_auc_score` functions.
+  You should call it and use as follows:
+
+      distinct_value_indices = Scholar.Metrics.calculate_distinct_value_indices(scores)
+      {fpr, tpr, thresholds} = Scholar.Metrics.roc_curve(y_true, scores, distinct_value_indices, weights)
+  """
+  def calculate_distinct_value_indices(y_score) do
+    desc_score_indices = Nx.argsort(y_score, direction: :desc)
+    y_score = Nx.take_along_axis(y_score, desc_score_indices)
+
+    distinct_value_indices_mask = Nx.not_equal(y_score[[1..-1//1]], y_score[[0..-2//1]])
+
+    Nx.iota({Nx.size(y_score) - 1})
+    |> Nx.add(1)
+    |> Nx.multiply(distinct_value_indices_mask)
+    |> Nx.to_flat_list()
+    |> Enum.filter(fn x -> x != 0 end)
+    |> Nx.tensor()
+    |> Nx.subtract(1)
```

Comment:
```suggestion
    |> Nx.multiply(distinct_value_indices_mask)
    |> Nx.subtract(-1)
    |> Nx.to_flat_list()
    |> Enum.filter(fn x -> x != -1 end)
    |> Nx.tensor()
```

Reply:
Shouldn't it be subtract 1?

Reply:
Sorry, fixed.

Reply:
I don't get this change

Reply:
I believe the idea is that if we keep the operations close together we maximize the chance of operation fusion at compile-time

Reply:
Now if first index in mask was set to `1` it will be filtered, but it shouldn't be

Reply:
We are filtering by -1 now. so if the mask is 1, we won't filter it. :)

Reply:
First element of iota is 0, 0 * 1 if is 0, and then when you subtract it will give -1, so you fillter it, but it shouldn't be filltered. To problem here is caused by the 0 from iota

Reply:
But don’t we do Nx.add(1)? I am struggling to see for which inputs the results would differ. :(

Reply:
Ok sorry, I completely forget about that add(1)